custom_components/xiaomi_home/miot/miot_lan.py

```diff
@@ -857,18 +866,68 @@ def get_device_list_handler(msg: dict, fut: asyncio.Future):
                 fut.set_result, msg)
 
         fut: asyncio.Future = self._main_loop.create_future()
-        if self.__lan_send_cmd(
-            MIoTLanCmdType.GET_DEV_LIST,
-            MIoTLanGetDevListData(
+        self._internal_loop.call_soon_threadsafe(
+            self.__get_dev_list,
+            _MIoTLanGetDevListData(
                 handler=get_device_list_handler,
                 handler_ctx=fut,
-                timeout_ms=timeout_ms)):
-            return await fut
-        _LOGGER.error('get_dev_list_async error, send cmd failed')
-        fut.set_result({})
+                timeout_ms=timeout_ms))
+        return await fut
+
+    async def __call_api_async(
+        self, did: str, msg: dict, timeout_ms: int = 10000
+    ) -> dict:
+        def call_api_handler(msg: dict, fut: asyncio.Future):
+            self._main_loop.call_soon_threadsafe(
+                fut.set_result, msg)
+
+        fut: asyncio.Future = self._main_loop.create_future()
+        self._internal_loop.call_soon_threadsafe(
+            self.__call_api, did, msg, call_api_handler, fut, timeout_ms)
         return await fut
 
-    def ping(self, if_name: str, target_ip: str) -> None:
+    async def __on_network_info_change_external_async(
+        self,
+        status: InterfaceStatus,
+        info: NetworkInfo
+    ) -> None:
+        _LOGGER.info(
+            'on network info change, status: %s, info: %s', status, info)
+        available_net_ifs = set()
+        for if_name in list(self._network.network_info.keys()):
+            available_net_ifs.add(if_name)
+        if len(available_net_ifs) == 0:
+            await self.deinit_async()
+            self._available_net_ifs = available_net_ifs
+            return
+        if self._net_ifs.isdisjoint(available_net_ifs):
+            _LOGGER.info('no valid net_ifs')
+            await self.deinit_async()
+            self._available_net_ifs = available_net_ifs
+            return
+        if not self._init_done:
+            self._available_net_ifs = available_net_ifs
+            await self.init_async()
+            return
+        self._internal_loop.call_soon_threadsafe(
+            self.__on_network_info_chnage,
+            _MIoTLanNetworkUpdateData(status=status, if_name=info.name))
+
+    async def __on_mips_service_change(
+        self, group_id: str,  state: MipsServiceState, data: dict
+    ) -> None:
+        _LOGGER.info(
+            'on mips service change, %s, %s, %s',  group_id, state, data)
+        if len(self._mips_service.get_services()) > 0:
+            _LOGGER.info('find central service, deinit miot lan')
+            await self.deinit_async()
+        else:
+            _LOGGER.info('no central service, init miot lan')
+            await self.init_async()
+
+# The folowing methods SHOULD ONLY be called in the internal loop
```

Comment:
spelling mistakes: folowing

Reply:
Fixed in 7b41773f63067a0839842e44ca01d624bb44856a